client-lite/src/download/download.cpp

```diff
@@ -432,20 +444,135 @@ UINT Download::_MaxNoProgressIntervals() const
+        return g_progressTrackerMaxNoProgressIntervals;
+    }
+
+    auto maxNoProgressIntervals = std::round(std::chrono::duration<double>(_noProgressTimeout) / g_progressTrackerCheckInterval);
+    DO_ASSERT(maxNoProgressIntervals >= 1);
+    return static_cast<UINT>(maxNoProgressIntervals);
+}
+
+std::string Download::_UpdateConnectionTypeAndGetUrl(bool retryAfterFailure)
+{
+    auto newConnectionType = _connectionType;
+    if (retryAfterFailure)
+    {
+        DO_ASSERT(_connectionType != ConnectionType::None);
+
+        _fAllowMcc = !_FallbackFromMccDue();
+
+        if (_fAllowMcc && !_mccFallbackDue)
+        {
+            // Use of MCC is allowed and fallback config not overriden.
+            // Use custom logic to decide when to switch between MCC and CDN.
+
+            const bool noProgressOnCurrentConnectionType = _cbTransferredAtRequestBegin == _status.BytesTransferred;
+            if (_connectionType == ConnectionType::MCC)
+            {
+                if (_fOriginalHostAttempted)
+                {
+                    if (noProgressOnCurrentConnectionType && (_numAttemptsWithCurrentConnectionType >= 2))
+                    {
+                        newConnectionType = ConnectionType::CDN;
+                    }
+                }
+                else
+                {
+                    newConnectionType = ConnectionType::CDN;
+                }
+            }
+            else if (_connectionType == ConnectionType::CDN)
+            {
+                if (noProgressOnCurrentConnectionType && (_numAttemptsWithCurrentConnectionType >= 2))
+                {
+                    newConnectionType = ConnectionType::MCC;
+                }
+            }
+            else
+            {
+                DO_ASSERT(false);
+            }
+        }
+    }
+    else
+    {
+        // Initial start or resume from caller or after transient error, always start with MCC
+        newConnectionType = ConnectionType::MCC;
+    }
+
+    DO_ASSERT(newConnectionType != ConnectionType::None);
+
+    std::string urlToUse;
+    if ((newConnectionType == ConnectionType::MCC) && _fAllowMcc && !_mccHost.empty() && !_mccManager.IsBanned(_mccHost, _url))
+    {
+        newConnectionType = ConnectionType::MCC;
+        urlToUse = SwapUrlHostNameForMCC(_url, _mccHost);
+    }
+    else
+    {
+        newConnectionType = ConnectionType::CDN;
+        urlToUse = _url;
+    }
+
+    if (newConnectionType != _connectionType)
+    {
+        _connectionType = newConnectionType;
+        _numAttemptsWithCurrentConnectionType = 1;
+    }
+    else
+    {
+        _numAttemptsWithCurrentConnectionType++;
+    }
+
+    DoLogInfo("%s, connection type: %d, numAttempts: %u, url: %s",
+        GuidToString(_id).data(), static_cast<int>(_connectionType), _numAttemptsWithCurrentConnectionType, urlToUse.data());
+    return urlToUse;
+}
+
+bool Download::_ShouldFallbackFromMccNow(bool isFatalError) const
```

Comment:
>_ShouldFallbackFromMccNow [](http://example.com/codeflow?start=15&length=25)

Name is kind of confusing, consider _ShouldStopUsingMCC

Reply:
It's not a permanent switch to CDN so I'll rename to _ShouldPauseMccUsage().